plugins/modules/packaging/os/pacman.py

```diff
@@ -180,357 +183,468 @@
-    boolean to indicate if the package is up-to-date and a third boolean to indicate whether online information were available
-    """
-
-    lcmd = "%s --query %s" % (pacman_path, name)
-    lrc, lstdout, lstderr = module.run_command(lcmd, check_rc=False)
-    if lrc != 0:
-        # package is not installed locally
-        return False, False, False
-    else:
-        # a non-zero exit code doesn't always mean the package is installed
-        # for example, if the package name queried is "provided" by another package
-        installed_name = get_name(module, lstdout)
-        if installed_name != name:
-            return False, False, False
-
-    # no need to check the repository if state is present or absent
-    # return False for package version check, because we didn't check it
-    if state == 'present' or state == 'absent':
-        return True, False, False
-
-    # get the version installed locally (if any)
-    lversion = get_version(lstdout)
-
-    rcmd = "%s --sync --print-format \"%%n %%v\" %s" % (pacman_path, name)
-    rrc, rstdout, rstderr = module.run_command(rcmd, check_rc=False)
-    # get the version in the repository
-    rversion = get_version(rstdout)
-
-    if rrc == 0:
-        # Return True to indicate that the package is installed locally, and the result of the version number comparison
-        # to determine if the package is up-to-date.
-        return True, (lversion == rversion), False
-
-    # package is installed but cannot fetch remote Version. Last True stands for the error
-    return True, True, True
-
-
-def update_package_db(module, pacman_path):
-    if module.params['force']:
-        module.params["update_cache_extra_args"] += " --refresh --refresh"
-
-    cmd = "%s --sync --refresh %s" % (pacman_path, module.params["update_cache_extra_args"])
-    rc, stdout, stderr = module.run_command(cmd, check_rc=False)
-
-    if rc == 0:
-        return stdout, stderr
-    else:
-        module.fail_json(msg="could not update package db", stdout=stdout, stderr=stderr)
-
-
-def upgrade(module, pacman_path):
-    cmdupgrade = "%s --sync --sysupgrade --quiet --noconfirm %s" % (pacman_path, module.params["upgrade_extra_args"])
-    cmdneedrefresh = "%s --query --upgrades" % (pacman_path)
-    rc, stdout, stderr = module.run_command(cmdneedrefresh, check_rc=False)
-    data = stdout.split('\n')
-    data.remove('')
-    packages = []
-    diff = {
-        'before': '',
-        'after': '',
-    }
-
-    if rc == 0:
-        # Match lines of `pacman -Qu` output of the form:
-        #   (package name) (before version-release) -> (after version-release)
-        # e.g., "ansible 2.7.1-1 -> 2.7.2-1"
-        regex = re.compile(r'([\w+\-.@]+) (\S+-\S+) -> (\S+-\S+)')
-        for p in data:
-            if '[ignored]' not in p:
-                m = regex.search(p)
-                packages.append(m.group(1))
-                if module._diff:
```

Comment:
Thanks to your changelog message I noticed that you did not implement diff mode in the rewite, i.e. you effectively removed diff support for the module in this PR. This is something thad you need to implement as well.

Reply:
Hmm, it looks like I was somehow blind this morning... (probably not fully awake...) Somehow I no longer managed to find the diff code in the new version of the module and managed to overlook it when testing... Sorry for my confusion!